src/wit_gen.rs

```diff
@@ -114,12 +114,44 @@ impl<F> TraceContext<F> {
 
 pub type Trace<F, TraceArgs> = dyn Fn(&mut TraceContext<F>, TraceArgs) + 'static;
 
-pub struct TraceGenerator<F, TraceArgs> {
-    trace: Rc<Trace<F, TraceArgs>>,
+/// A simple trace generator that generates a trace witness by calling the trace function with the
+/// given arguments.
+///
+/// ### Parameters
+/// - `F`: The field type.
+/// - `TA`: The type of arguments that the trace function takes.
+///
+/// ### Fields
+/// - `trace`: The trace function.
+/// - `num_steps`: The number of steps in the circuit.
+pub struct SimpleTraceGenerator<F, TA = ()> {
+    trace: Rc<Trace<F, TA>>,
     num_steps: usize,
 }
 
-impl<F, TraceArgs> Clone for TraceGenerator<F, TraceArgs> {
+/// A trait that defines the interface for trace generators. A trace generator is responsible for
+/// generating a trace witness for the given arguments.
+///
+/// ### Parameters
+/// - `F`: The field type.
+///
+/// ### Associated Types
+/// - `TraceGenerator::TraceArgs` The type of arguments that the trace generator takes.
+pub trait TraceGenerator<F> {
+    /// The type of arguments that the trace generator takes.
+    type TraceArgs: Clone;
+    /// Creates a new trace generator with the given trace function and number of steps.
+    /// ### Parameters
+    /// - `trace`: The trace function.
+    /// - `num_steps`: The number of steps in the circuit.
+    fn new(trace: Rc<Trace<F, Self::TraceArgs>>, num_steps: usize) -> Self;
```

Comment:
So we are trying that different implementations could have a different way of generating TraceWitness, and if that case their constructor would be different

Reply:
Please, check this @alxkzmn I don't think new should be in the trait

Reply:
https://github.com/privacy-scaling-explorations/chiquito/pull/257#discussion_r1625666432